sw/device/lib/dif/dif_clkmgr.c

```diff
@@ -3,7 +3,131 @@
 // SPDX-License-Identifier: Apache-2.0
 
 #include "sw/device/lib/dif/dif_clkmgr.h"
+#include "sw/device/lib/base/bitfield.h"
+#include "sw/device/lib/base/mmio.h"
 
 #include "clkmgr_regs.h"  // Generated
 
-#error "Clock Manager DIF Implementation Incomplete"
+/**
+ * There are fewer Gateable clocks than this in the hardware, but this is the
+ * threshold at which we will need to change the implementation of the DIF.
+ */
+static const uint32_t kMaxGateableClocks = CLKMGR_PARAM_REG_WIDTH;
```

Comment:
Are you sure you are using this define correctly? This just tells you how wide the registers are. I am not sure I understand how these are relatable to `kMaxGateableClocks` and `kMaxHintableClocks`.

I think `kMaxGateableClocks` and `kMaxHintableClocks` should be passed into the DIF as `dif_clkmgr_params_t params` to the `init` function (with values taken from the relevant `top_earlgrey` definitions).

Reply:
We're waiting on https://github.com/lowRISC/opentitan/issues/3932 to get these values at all - until then, assuming 32 is not bad, as we know we maximum have one register of them. 

In the long run we probably do want these in `params_t`, but so far we haven't required that DIFs are fully parameterisable, just that we want to get there eventually.

Reply:
I took the other approach of adding `last_hintable_clock` and `last_gateable_clock` to `dif_clkmgr_params_t`.